ios/ObjectPermissions/ToDo/ProjectsViewController.swift

```diff
@@ -0,0 +1,193 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+import UIKit
+import RealmSwift
+
+class ProjectsViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {
+    
+    let realm: Realm
+    let projects: Results<Project>
+    var notificationToken: NotificationToken?
+    var subscriptionToken: NotificationToken?
+    var subscription: SyncSubscription<Project>!
+    
+    var tableView = UITableView()
+    let activityIndicator = UIActivityIndicatorView()
+    
+    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
+        let syncConfig = SyncConfiguration(user: SyncUser.current!, realmURL: Constants.REALM_URL, isPartial: true)
+        realm = try! Realm(configuration: Realm.Configuration(syncConfiguration: syncConfig))
+
+        // Display all projects that the user has permissions to see.
+        projects = realm.objects(Project.self).sorted(byKeyPath: "timestamp", ascending: false)
+        
+        super.init(nibName: nil, bundle: nil)
+    }
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        title = "My Projects"
+        view.addSubview(tableView)
+        view.addSubview(activityIndicator)
+        activityIndicator.center = self.view.center
+        activityIndicator.color = .darkGray
+        activityIndicator.isHidden = false
+        activityIndicator.hidesWhenStopped = true
+        
+        tableView.frame = self.view.frame
+        tableView.delegate = self
+        tableView.dataSource = self
+        
+        navigationItem.leftBarButtonItem = UIBarButtonItem(barButtonSystemItem: .add, target: self, action: #selector(addItemButtonDidClick))
+        navigationItem.rightBarButtonItem = UIBarButtonItem(title: "Logout", style: .plain, target: self, action: #selector(logoutButtonDidClick))
+        
+        // In a Partial Sync use case this is where we tell the server we want to
+        // subscribe to a particular query.
+        subscription = projects.subscribe(named: "my-projects")
+        
+        activityIndicator.startAnimating()
+        subscriptionToken = subscription.observe(\.state, options: .initial) { state in
+            print("Subscription State: \(state)")
+            if state == .complete {
+                self.activityIndicator.stopAnimating()
+            }
+        }
```

Comment:
How does this differ from the `observe` below? Is it a different kind of `.initial` state?

Reply:
I am still not clear how this is different. Won't this also trigger `.initial` in the collection notification? Why do we need both?

Reply:
No, the `.initial` notification is fired as soon as the query is executed for the first time. The query having been executed tells you nothing about the state of any sync subscription, if one even exists, since the query is executed against data in the local Realm. When the subscription is processed by the server and it begins syncing matching objects to the client, it will quite likely result in the contents of the collection changing. This will result in a `.update` notification that describes the changes.

We need both because they tell you completely different things.

Reply:
_Note, the point of the beta period is to gather early API feedback before we go GA and lock it down. This is just me starting that process based on looking at actual code and playing around with the API myself. This is meant for discussion and not necessarily something we need to act on now._

I get that you can have two kinds of initial state:

1. Result of first query against local state. This can contain pretty much anything:
    * nothing if this it the first query and nothing has been synced yet.
    * random subset of objects if other queries randomly happen to have pulled in objects that match this query.
    * the full result set if you have previously done this query and the result is still cached).

2. Full and correct query result from the server.

1 seems like it has the potential to be very confusing to the user (given that it is very hard to predict what it will actually contain). I wonder if there are ways to make it more clear.

The whole sequence of getting to the query seems a bit cumbersome. Having to maintain 4 variables (the query, the subscription and two tokens), plus all the boilerplate code to set them all up feels very heavy. We should keep trying to find a way to roll them together.

Reply:
Following up on the above. We really need a 3rd state:

3. **Result from server or cache of last result**

So basically the intersection of the two other states, but without including random state from local state from before you have gotten any result from the server.

It should be possible to look at the subscription object, and only return the event if it has been updated by the server at least once. That would preclude returning random state, but would still allow it to return a locally cached state as long as the subscription has been filled as least once.

To me that seems like the state you would want 99% of the time, and it is actually pretty hard to discern based on the states currently exposed (how do you differentiate full state from cache from incomplete?).